### PR TITLE
JIT support input size changing

### PR DESCRIPTION
**Pull request by amatraturi**
Added support for input size changing (as per https://github.com/geohot/tinygrad/issues/692)
Added Test for size changing jit (passes for this PR)
Let me know if this is proper.

### CODE REVIEW

**Comment by geohot** [comment]
Code. No tests.

Does this add new behavior? Why isn't the new behavior tested?

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/jit.py                           24              14
------------------------------------------------------------
lines added in the tinygrad folder: 10
```

**Comment by amatraturi** [comment]
> If you add code, you have to add tests.
> 
> Does this add new behavior? The new behavior needs to be tested.

Yes, understood. Added the test for input size changing, passes with this PR.

Existing usage (same size) would behave same as previously.

File: tinygrad/jit.py [hunk 1]
```diff
@@ -19,33 +19,41 @@ def __get__(self, obj, objtype): return functools.partial(self.__call__, obj)
 
   def __call__(self, *args, **kwargs) -> Any:
     if Device.DEFAULT not in ["GPU", "CLANG", "METAL", "CUDA"]: return self.fxn(*args, **kwargs)  # only jit on the GPU codegen
+
     # NOTE: this cast is needed since although we know realize will create a ".realized" DeviceBuffer, the type checker doesn't
     input_rawbuffers: Dict[Union[int, str], RawBuffer] = {cast(Union[int, str], k):cast(RawBuffer, v.realize().lazydata.realized) for k,v in itertools.chain(enumerate(args), kwargs.items()) if isinstance(v, Tensor)}
+    specialised_key = f'{hash("".join([str(input_rawbuffers[i].size) for i,k in input_rawbuffers.items()]))}'
```
Comment (geohot) [inline tinygrad/jit.py:25]: What if the shapetrackers are different? Like permuted? Is just checking the size enough?

File: test/test_jit.py [hunk 2]
```diff
@@ -70,5 +70,16 @@ def __call__(self, b:Tensor) -> Tensor:
       c = fun(b)
       np.testing.assert_equal(c.numpy(), fun.a.numpy()+b.numpy())
 
+  def test_shape_change_jit(self):
```
Comment (geohot) [inline test/test_jit.py:73]: I will merge just this test with an expected failure for now if you want to submit it.

**Pull request closed without merging**